src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -817,209 +817,380 @@ private BoundCollectionExpression ConvertCollectionExpression(
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private readonly BoundCollectionExpression? TryConvertCollectionExpressionArrayOrSpanType(
+                CollectionExpressionTypeKind collectionTypeKind,
+                ImmutableArray<BoundNode> elements)
             {
-                if (targetType is NamedTypeSymbol namedType &&
-                    HasParamsCollectionTypeInProgress(namedType, out NamedTypeSymbol? inProgress, out MethodSymbol? inProgressConstructor))
+                var syntax = _node.Syntax;
+                switch (collectionTypeKind)
+                {
+                    case CollectionExpressionTypeKind.Span:
+                        _ = _binder.GetWellKnownTypeMember(WellKnownMember.System_Span_T__ctor_Array, _diagnostics, syntax: syntax);
+                        break;
+
+                    case CollectionExpressionTypeKind.ReadOnlySpan:
+                        _ = _binder.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Array, _diagnostics, syntax: syntax);
+                        break;
+                }
+
+                if (_node.WithElement != null)
+                {
+                    // Strictly disallowed in the language specification for: Array, Span, ReadOnlySpan
+                    _diagnostics.Add(
+                        ErrorCode.ERR_CollectionArgumentsNotSupportedForType,
+                        _node.WithElement.Syntax.GetFirstToken().GetLocation(),
+                        _targetType);
+                }
+
+                return new BoundCollectionExpression(
+                    syntax,
+                    collectionTypeKind,
+                    placeholder: null,
+                    collectionCreation: null,
+                    collectionBuilderMethod: null,
+                    collectionBuilderElementsPlaceholder: null,
+                    wasTargetTyped: true,
+                    // Since a with-element is not legal here, we can just treat these as if they don't have one at all.
+                    hasWithElement: false,
+                    _node,
+                    elements,
+                    _targetType);
+            }
+
+            private readonly BoundCollectionExpression? TryConvertCollectionExpressionArrayInterfaceType(ImmutableArray<BoundNode> elements)
+            {
+                var syntax = _node.Syntax;
+
+                var collectionCreation = BindCollectionArrayInterfaceConstruction();
+
+                return new BoundCollectionExpression(
+                    syntax,
+                    CollectionExpressionTypeKind.ArrayInterface,
+                    placeholder: null,
+                    collectionCreation,
+                    collectionBuilderMethod: null,
+                    collectionBuilderElementsPlaceholder: null,
+                    wasTargetTyped: true,
+                    // Since a with-element is not legal here, we can just treat these as if they don't have one at all.
```

Comment:
>// Since a with-element is not legal here, we can just treat these as if they don't have one at all. [](http://example.com/codeflow?start=20&length=100)

This doesn't look accurate

Reply:
dfdd386167c